llvm/lib/Target/AMDGPU/AMDGPULowerKernelArguments.cpp

```diff
@@ -133,26 +134,26 @@ static void addAliasScopeMetadata(Function &F, DataLayout const &DL) {
       continue;
 
     // Collect underlying objects of pointer arguments.
-    auto Scopes = SmallVector<Metadata *, 4u>();
-    auto ObjSet = SmallPtrSet<Value const *, 4u>();
-    auto NoAliases = SmallVector<Metadata *, 4u>();
+    SmallVector<Metadata *, 4u> Scopes;
+    SmallPtrSet<Value const *, 4u> ObjSet;
+    SmallVector<Metadata *, 4u> NoAliases;
 
-    for (auto &Ptr : PtrArgs) {
-      auto Objects = SmallVector<Value const *, 4u>();
+    for (Value const *&Ptr : PtrArgs) {
+      SmallVector<Value const *, 4u> Objects;
       getUnderlyingObjects(Ptr, Objects);
       ObjSet.insert_range(Objects);
     }
 
-    auto RequiresNoCaptureBefore = false;
-    auto UsesUnknownObject = false;
-    auto UsesAliasingPtr = false;
+    bool RequiresNoCaptureBefore = false;
+    bool UsesUnknownObject = false;
+    bool UsesAliasingPtr = false;
 
-    for (auto *Val : ObjSet) {
+    for (Value const *Val : ObjSet) {
       if (isa<ConstantPointerNull>(Val) || isa<ConstantDataVector>(Val) ||
           isa<ConstantInt>(Val) || isa<ConstantFP>(Val) || isa<UndefValue>(Val))
         continue;
 
-      if (auto *Arg = dyn_cast<Argument>(Val)) {
+      if (Argument const *Arg = dyn_cast<Argument>(Val)) {
```

Comment:
```suggestion
      if (const Argument *Arg = dyn_cast<Argument>(Val)) {
```
The autos would also be OK in the cases where the type is present in the cast